### PR TITLE
exec replace transaction with write batch

### PR DESCRIPTION
**Pull request by liubaohai**
# Description           

exec 采用write batch代替事务

# Checklist:

- [x] **Self-review**: I have performed a **self-review** of my own code
- [ ] **Comment**:     I have **commented my code**, particularly in hard-to-understand areas
- [ ] **Doc**:         I have made corresponding changes to the **documentation**
- [ ] **No-warnings**: My changes generate **no new warnings**
- [x] **Add-test**:    I have added **tests** that prove my fix is effective or that my feature works
- [x] **Pass**:        New and existing **unit tests pass** locally with my changes

### CODE REVIEW

File: components/epaxos/src/replica/errors.rs [hunk 1]
```diff
@@ -20,6 +21,11 @@ quick_error! {
         Protocol(e: ProtocolError) {
             from(e: ProtocolError) -> (e)
         }
+
+        ProtoEncodeError(err: EncodeError) {
```
Comment (drmingdrmer) [inline components/epaxos/src/replica/errors.rs]:
```suggestion
        Prost(err: EncodeError) {
```
这样咋样, 直接能看出来是什么错误.

File: components/epaxos/src/snapshot/traits/traits.rs [hunk 2]
```diff
@@ -42,10 +44,57 @@ impl From<DBColumnFamily> for &str {
...
+    Set(DBColumnFamily, Vec<u8>, Vec<u8>),
+    Delete(DBColumnFamily, Vec<u8>),
+}
+
+impl From<&QCmd> for Command {
+    fn from(c: &QCmd) -> Self {
+        if OpCode::NoOp as i32 == c.op {
+            return Command::Nil;
+        } else if OpCode::Get as i32 == c.op {
+            return Command::Get(DBColumnFamily::Default, c.key.clone());
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/traits/traits.rs:59]: 看execute_commands的代码Get都已经先被执行放到repl里了, 貌似不用处理Get?

```
	                } else if cmd.op == OpCode::Get as i32 {
                    if !existed.contains_key(&cmd.key) {
                        let v = self.storage.get_kv(&cmd.key)?;
                        existed.insert(&cmd.key, v.clone());
                    }
                    repl.push(ExecuteResult::SuccessWithVal {
                        value: existed[&cmd.key].clone(),
                    });
```

File: components/epaxos/src/snapshot/traits/traits.rs [hunk 3]
```diff
@@ -42,10 +44,57 @@ impl From<DBColumnFamily> for &str {
...
-pub enum Command<'a> {
-    Get(DBColumnFamily, &'a Vec<u8>),
-    Set(DBColumnFamily, &'a Vec<u8>, &'a Vec<u8>),
-    Delete(DBColumnFamily, &'a Vec<u8>),
+pub enum Command {
+    Nil,
+    Get(DBColumnFamily, Vec<u8>),
+    Set(DBColumnFamily, Vec<u8>, Vec<u8>),
+    Delete(DBColumnFamily, Vec<u8>),
+}
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/traits/traits.rs:52]: 这个enum, 叫WriteEntry咋样? 给writebatch用的一个条目

File: components/epaxos/src/snapshot/traits/traits.rs [hunk 4]
```diff
@@ -42,10 +44,57 @@ impl From<DBColumnFamily> for &str {
...
+            return Command::Nil;
+        } else if OpCode::Get as i32 == c.op {
+            return Command::Get(DBColumnFamily::Default, c.key.clone());
+        } else if OpCode::Set as i32 == c.op {
+            return Command::Set(DBColumnFamily::Default, c.key.clone(), c.value.clone());
+        } else if OpCode::Delete as i32 == c.op {
+            return Command::Delete(DBColumnFamily::Default, c.key.clone());
+        } else {
+            panic!("unexpected operation: {:?}", c.op);
+        }
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/traits/traits.rs:66]: 对enum的匹配用match貌似就不用panic了.

File: components/epaxos/src/snapshot/traits/traits.rs [hunk 5]
```diff
@@ -42,10 +44,57 @@ impl From<DBColumnFamily> for &str {
...
+    }
+}
+
+impl From<Instance> for Command {
+    fn from(inst: Instance) -> Self {
+        let mut v = vec![];
+        inst.encode(&mut v).unwrap();
+        return Command::Set(
+            DBColumnFamily::Instance,
+            inst.instance_id.unwrap().to_key(),
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/traits/traits.rs:76]: inst有实现ToKey

File: components/epaxos/src/snapshot/traits/traits.rs [hunk 6]
```diff
@@ -42,10 +44,57 @@ impl From<DBColumnFamily> for &str {
...
+            v,
+        );
+    }
+}
+
+impl From<(&str, InstanceId)> for Command {
+    fn from(args: (&str, InstanceId)) -> Self {
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/traits/traits.rs:83]: 这样好像通用性更好点:
```
impl<A: Into<InstanceId> + Copy> From<(&str, A)> for Command {
```

File: components/epaxos/src/replica/errors.rs (hunk 1 above)
Comment (liubaohai) [reply components/epaxos/src/replica/errors.rs]: 不需要这个，我忘记删除了

File: components/epaxos/src/snapshot/traits/traits.rs (hunk 6 above)
Comment (liubaohai) [reply components/epaxos/src/snapshot/traits/traits.rs:83]: 厉害

File: components/epaxos/src/snapshot/traits/traits.rs (hunk 3 above)
Comment (liubaohai) [reply components/epaxos/src/snapshot/traits/traits.rs:52]: 当时没想到好的名字。WriteEntry不错，用这个吧

File: components/epaxos/src/snapshot/traits/traits.rs (hunk 6 above)
Comment (drmingdrmer) [reply components/epaxos/src/snapshot/traits/traits.rs]: `o(*////▽////*)q`

**Pull request merged**